### PR TITLE
feat: add chaos testing toolkit

### PR DESCRIPTION
**Pull request by WenyXu**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

- Add the `Cluster` trait, which provides the ability to deploy/destroy a cluster
  - Implement `DebugCluster`, which only lists all GreptimeDB nodes on the local machine (required to deploy the cluster manually).
- Add the `bare::client::Client`, which interacts with `Chaosd` server.
- Add the `Nemesis` trait, which injects failure and disturbs the database.
  - Implement `AttackProcess`, a process nemesis, which sends signals (SigTerm, SigKill, SigStop) to a specific process.

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

### CODE REVIEW

File: tests/chaos/src/chaos/bare/client/process.rs [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+const ATTACK_PROCESS_ENDPOINT: &str = "/api/attack/process";
+
+/// Signal Currently, only SIGKILL, SIGTERM, and SIGSTOP are supported.
+#[derive(Serialize_repr, Deserialize_repr, PartialEq, Debug, Clone, Copy)]
+#[repr(u8)]
+pub enum Signal {
+    Kill = 9,
+    Term = 15,
+    Stop = 23,
```
Comment (MichaelScofield) [inline tests/chaos/src/chaos/bare/client/process.rs:29]: better make these magic numbers const

File: tests/chaos/src/chaos/bare/nemesis/process.rs [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+use crate::chaos::bare::client::ClientRef;
+use crate::chaos::nemesis::Nemesis;
+use crate::cluster::bare::node::Node;
+use crate::error::{self, Result};
+
+#[derive(Debug)]
+pub struct AttackProcess {
+    signal: Signal,
+    client_ref: ClientRef,
+    // Returned by chaosd, used to recovers the attack.
```
Comment (MichaelScofield) [inline tests/chaos/src/chaos/bare/nemesis/process.rs:27]:
```suggestion
    // Returned by chaosd, used to recover from the attack.
```

File: tests/chaos/src/cluster/bare/debug.rs [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        Ok(())
+    }
+
+    async fn delete(&self) -> Result<()> {
+        Ok(())
+    }
+
+    async fn nodes(&self) -> Result<Vec<Node>> {
+        let processes = list_processes_by_name(GREPTIME_DB_EXEC_FILENAME.to_string())?;
+        // TODO(weny): add etcd nodes?
```
Comment (MichaelScofield) [inline tests/chaos/src/cluster/bare/debug.rs:49]: we need etcd to chaos testing the leader election of metasrv

File: tests/chaos/src/cluster/bare/process.rs [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+        let tcpnet_entries = process_ref.tcp().context(error::GetProcInfoSnafu)?;
+        let fds = process_ref.fd().context(error::GetProcInfoSnafu)?;
+        let inodes = fds
+            .filter_map(|fd| {
+                if let FDTarget::Socket(inode) = fd.unwrap().target {
+                    Some(inode)
+                } else {
+                    None
+                }
+            })
```
Comment (MichaelScofield) [inline tests/chaos/src/cluster/bare/process.rs:59]: can it:

```suggestion
            .filter_map(|fd| {
                fd.and_then(|x| 
                    if let FDTarget::Socket(inode) = x.target { Some(inode) } else { None }
                )
            })
```

File: tests/chaos/src/cluster/bare/debug.rs [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+fn parse_node_id(val: &str) -> u64 {
+    let node_id = DATANODE_NODE_ID.captures(val).unwrap()[1].to_string();
+    node_id.parse().unwrap()
+}
+
+fn new_debug_node(process: ProcessInfo) -> Node {
+    let info = match process.cmdline[1].as_str() {
+        "datanode" => {
+            // ["greptime", "datanode", "start", "--node-id=0", "--metasrv-addr=0.0.0.0:3002", "--rpc-addr=0.0.0.0:14100", "--http-addr=0.0.0.0:14300"]
+            let node_id = parse_node_id(&process.cmdline[3]);
```
Comment (MichaelScofield) [inline tests/chaos/src/cluster/bare/debug.rs:69]: Well, we can't simply get node id like this. The `greptime` could start with just `-c=config.toml`. Can we pass these info from outside?

File: tests/chaos/src/cluster/bare/process.rs [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+            let procs = procs
+                .into_iter()
+                .filter_map(|proc| match ProcessInfo::try_new(proc) {
+                    Ok(proc) => Some(proc),
+                    Err(_) => None,
+                })
+                .collect::<Vec<_>>();
+
+            for proc in procs {
+                // println!("{:?}", proc);
```
Comment (MichaelScofield) [inline tests/chaos/src/cluster/bare/process.rs:120]:
```suggestion
```

File: tests/chaos/src/chaos/bare/client/process.rs [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+#[derive(Debug, Serialize, Deserialize)]
+pub struct AttackProcessRequest {
+    process: String,
+    signal: Signal,
+}
+
+impl Client {
+    pub async fn attack_process(&self, process: i32, sig: Signal) -> Result<Response> {
+        let req = AttackProcessRequest {
+            process: format!("{}", process),
```
Comment (MichaelScofield) [inline tests/chaos/src/chaos/bare/client/process.rs:41]: Why not just make `process` field an `i32`?

**Review by MichaelScofield — COMMENTED** [review]
please add more docs about the design of your chaos testing integration

**Review by MichaelScofield — COMMENTED** [review]
please add more docs about the design of your chaos testing integration

File: tests/chaos/src/chaos/bare/client/process.rs (hunk 1 above)
Comment (WenyXu) [reply tests/chaos/src/chaos/bare/client/process.rs:29]: > better make these magic numbers const

The `chaosd` server accepts a String type `process` field. It also accepts to attack a process by the process's name. However, we only use the process id here.

File: tests/chaos/src/chaos/bare/client/process.rs (hunk 7 above)
Comment (WenyXu) [reply tests/chaos/src/chaos/bare/client/process.rs:41]: > Why not just make `process` field an `i32`?

The `chaosd` server accepts a String type `process` field. It also accepts to attack a process by the process's name. However, we only use the process id here.

File: tests/chaos/src/cluster/bare/debug.rs (hunk 3 above)
Comment (WenyXu) [reply tests/chaos/src/cluster/bare/debug.rs:49]: > we need etcd to chaos testing the leader election of metasrv

I will add the feature in the future PR.

**Pull request closed without merging**